strawberryfields/apps/train/parametrization.py

```diff
@@ -0,0 +1,406 @@
+# limitations under the License.
+r"""
+Contains the :class:`~.VGBS` (variational GBS) class which provides a trainable parametrization
+of the GBS probability distribution.
+"""
+from typing import Optional
+
+import numpy as np
+import thewalrus.samples
+from thewalrus._hafnian import reduction
+from thewalrus._torontonian import tor
+from thewalrus.quantum import Qmat, Xmat
+from thewalrus.quantum import find_scaling_adjacency_matrix as rescale
+from thewalrus.quantum import find_scaling_adjacency_matrix_torontonian as rescale_tor
+from thewalrus.quantum import photon_number_mean_vector, pure_state_amplitude
+
+import strawberryfields as sf
+
+
+def rescale_adjacency(A: np.ndarray, n_mean: float, threshold: bool) -> np.ndarray:
+    """Rescale an adjacency matrix so that it can be mapped to GBS.
+
+    An adjacency matrix must have singular values not exceeding one if it can be mapped to GBS.
+    Arbitrary adjacency matrices must first be rescaled to satisfy this condition.
+
+    This function rescales an input adjacency matrix :math:`A` so that the corresponding gaussian
+    state has:
+
+    - a mean number of *clicks* equal to ``n_mean`` when ``threshold=True``;
+    - a mean number of *photons* equal to ``n_mean`` when ``threshold=False``.
+
+    **Example usage:**
+
+    >>> a = np.ones((3, 3))
+    >>> rescale_adjacency(a, 2, True)
+    array([[0.32232919, 0.32232919, 0.32232919],
+           [0.32232919, 0.32232919, 0.32232919],
+           [0.32232919, 0.32232919, 0.32232919]])
+
+    Args:
+        A (array): the adjacency matrix to rescale
+        n_mean (float): the target mean number of clicks or mean number of photons
+        threshold (bool): determines whether rescaling is for a target mean number of clicks or
+            photons
+
+    Returns:
+        array: the rescaled adjacency matrix
+    """
+    scale = rescale_tor(A, n_mean) if threshold else rescale(A, n_mean)
+    return A * scale
+
+
+def _Omat(A: np.ndarray) -> np.ndarray:
+    """Find the :math:`O` matrix of an adjacency matrix.
+
+    Args:
+        A (array): the adjacency matrix
+
+    Returns:
+        array: the :math:`O` matrix of :math:`A`
+    """
+    n = len(A)
+    A_big = np.block([[A, 0 * A], [0 * A, np.conj(A)]])
+    X = Xmat(n)
+    return X @ A_big
+
+
+def prob_click(A: np.ndarray, sample: np.ndarray):
+    """Calculate the probability of a click pattern.
+
+    The input adjacency matrix must have singular values not exceeding one. This can be achieved
+    by rescaling an arbitrary adjacency matrix using :func:`rescale_adjacency`.
+
+    Args:
+        A (array): the adjacency matrix
+        sample (array): the sample
+
+    Returns:
+        float: the probability
+    """
+    n = len(A)
+    O = _Omat(A)
+    sample_big = np.hstack([sample, sample]).astype("int")
+    O_sub = reduction(O, sample_big)
+    scale = np.sqrt(np.linalg.det(np.identity(2 * n) - O))
+    return scale * tor(O_sub)
```

Comment:
Could be useful to include something similar to [this one](https://the-walrus.readthedocs.io/en/latest/gbs.html#gaussian-states-and-threshold-detection) at a later stage as part of an explanation here

Reply:
Thanks! Good point, we'll revisit the docstrings again once all the features are in.